### ISSUE TITLE
CassandraDatacenter updates fail due to resource version conflicts

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#80]
While doing some manual testing I observed errors in the operator log when k8ssandra-operator was attempting to update CassandraDatacenters.

Here is the manifest for my K8ssandraCluster:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  name: test
spec:
  cassandra:
    cluster: test
    datacenters:
      - metadata:
          name: dc1
        k8sContext: jsanda-1
        size: 3
        serverVersion: "3.11.10"
        storageConfig:
          cassandraDataVolumeClaimSpec:
            storageClassName: standard
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 5Gi
        config:
          jvm-options:
            initial_heap_size: "1024m"
            max_heap_size: "1024m"
        resources:
          limits:
             memory: 2048Mi
        stargate:
          size: 1
          heapSize: 512Mi
          resources:
            limits:
              memory: 1024Mi
      - metadata:
          name: dc2
        k8sContext: jsanda-2
        size: 3
        serverVersion: "3.11.10"
        storageConfig:
          cassandraDataVolumeClaimSpec:
            storageClassName: standard
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 5Gi
        config:
          jvm-options:
            initial_heap_size: "1024m"
            max_heap_size: "1024m"
        resources:
          limits:
            memory: 2048Mi
        stargate:
          size: 1
          heapSize: 512Mi
          resources:
            limits:
              memory: 1024Mi
      - metadata:
          name: dc3
        k8sContext: jsanda-3
        size: 3
        serverVersion: "3.11.10"
        storageConfig:
          cassandraDataVolumeClaimSpec:
            storageClassName: standard
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 5Gi
        config:
          jvm-options:
            initial_heap_size: "1024m"
            max_heap_size: "1024m"
        resources:
          limits:
            memory: 2048Mi
        stargate:
          size: 1
          heapSize: 512Mi
          resources:
            limits:
              memory: 512Mi      
```

The errors look like this:

```
2021-08-12T21:38:33.668Z        ERROR   controller-runtime.manager.controller.k8ssandracluster  Failed to update datacenter 
 {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test", "namespace": "default", 
"CassandraDatacenter": {"namespace": "default", "name": "dc2"}, "error": "Operation cannot be fulfilled on
cassandradatacenters.cassandra.datastax.com \"dc2\": the object has been modified; please apply your changes to the latest
version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
github.com/k8ssandra/k8ssandra-operator/controllers.(*K8ssandraClusterReconciler).reconcile
        /workspace/controllers/k8ssandracluster_controller.go:137
github.com/k8ssandra/k8ssandra-operator/controllers.(*K8ssandraClusterReconciler).Reconcile
        /workspace/controllers/k8ssandracluster_controller.go:76
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
```

I made no update to the K8ssandraCluster spec but had lots of errors like this in the log.

While a resource version conflict is often perfectly normal, I think this deserves some investigation. We perform an update when the hash of the desired spec does not match the spec of the actual spec. The only update that the operator performs right now on its own without changes to the K8ssandraCluster spec is to update `AdditionalSeeds`.

We compute the hash after the computing the remote seeds endpoints. That should be excluded from computing the resource hash as it won't trigger a StatefulSet update.

Lastly, we should think about adding some retry logic since we have both k8ssandra-operator and cass-operator updating the CassandraDatacenter. The former updates the spec and metadata while the latter update the status.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-795) by [Unito](https://www.unito.io)
┆Fix Versions: k8ssandra-2.0.0-alpha
┆Issue Number: K8SSAND-795
┆Priority: Medium

**Comment by Miles-Garnsey** [comment]
_(Notes for myself on what is happening here)_ 

This error appears only when the cluster is first created and is in the process of bootstrapping before it moves to a ready state. I'm not seeing this error cropping up when making changes to the `k8ssandraclusters` (e.g. change some java heap settings and the changes propagate through without conflicts arising).

I see this error on a single DC cluster, so it isn't like we're running multiple controllers or anything in different DCs.

It seems to be triggered by [line 110](https://github.com/k8ssandra/k8ssandra-operator/blob/d8abaf59e906e22f28c39b43cf6bf58951db8ec2/controllers/k8ssandracluster_controller.go#L110) in the k8ssandra controller, which is concerned with setting the status on the DCs. The use of a `MergeFromWithOptimisticLock` in line [384](https://github.com/k8ssandra/k8ssandra-operator/blob/d8abaf59e906e22f28c39b43cf6bf58951db8ec2/controllers/k8ssandracluster_controller.go#L384) is what generates the inner error.

Will examine in more detail tomorrow.

**Comment by jdonenine** [comment]
@burmanm suggested in another ticket (Not sure where exactly, but we'll add back for reference) that we use the seed service to provide this info to cass-operator, so that we can completely avoid the update of the configuration.

**Comment by burmanm** [comment]
https://github.com/k8ssandra/k8ssandra-operator/issues/67

**Comment by jsanda** [comment]
The only change made to the CassandraDatacenter spec is adding/updating the `AdditionalSeeds` property. Part of the problem in this ticket that I want to update is the frequent updates that are happening. 

I modified the logging statement in k8ssandracluster_controller.go as follows:

```go
if actualHash, found := actualDc.Annotations[api.ResourceHashAnnotation]; !(found && actualHash == desiredDcHash) {
    logger.Info("Updating datacenter", "CassandraDatacenter", dcKey, "desiredDc", desiredDc.Spec, "actualDc", actualDc.Spec)
    actualDc = actualDc.DeepCopy()
    resourceVersion := actualDc.GetResourceVersion()
    desiredDc.DeepCopyInto(actualDc)
    actualDc.SetResourceVersion(resourceVersion)
    if err = remoteClient.Update(ctx, actualDc); err != nil {
        logger.Error(err, "Failed to update datacenter", "CassandraDatacenter", dcKey)
        return ctrl.Result{}, err
    }
}
```
I took the logs and diff a json diff which revealed the issue:

![Screen Shot 2021-08-19 at 1 38 29 PM](https://user-images.githubusercontent.com/168121/130118491-25e43616-fc93-48eb-9ef4-58fb1e2b4f31.png)
 
The order of elements in `AdditionalSeeds` is changing which in turn is causing the resource hash to change. This is unfortunate it. In light of this I think a strong argument could be made that `AdditionalSeeds` should be map and not a slice; however, I don't think we can easily make that change in cass-operator as it would definitely break backwards compatibility. 

For now, we need to make sure the order of elements in `AdditionalSeeds` does not change.

**Comment by Miles-Garnsey** [comment]
That's weird.=, I thought slices [did](https://stackoverflow.com/questions/45346641/slice-iteration-order-in-go) keep order consistent? Either way I guess we can just convert additionalSeeds into a set for comparison purposes and not update it if the current + desired sets are equal I guess.

**Comment by jsanda** [comment]
The ordering of slices is consistent. The `seeds` variable get populated during each reconciliation loop from a query against the api server. There is no ordering guarantee with the query results, hence the problem. I have updated the function that performs the query to sort the results so that we have consistent ordering. With that change I am no longer seeing this problem.

**Comment by jsanda** [comment]
Fixed in #73
